toolchain/sem_ir/typed_insts.h

```diff
@@ -821,6 +821,22 @@ struct ReturnExpr {
   InstId dest_id;
 };
 
+// Given an instruction with a constant value that depends on a generic
+// parameter, selects an instance of that instruction with the constant value
+// corresponding to a particular specific.
+//
+// TODO: We only form these as the instruction referenced by a `NameRef`.
+// Consider merging an `SpecificConstant` + `NameRef` into a new form of
+// instruction in order to give a more compact representation.
+struct SpecificConstant {
+  static constexpr auto Kind = InstKind::SpecificConstant.Define<Parse::NodeId>(
+      {.ir_name = "specific_constant", .is_lowered = false});
+
+  TypeId type_id;
+  InstId inst_id;
+  GenericInstanceId instance_id;
```

Comment:
Should `GenericInstanceId instance_id` -> `SpecificId specific_id`? Sort of a broad comment here. But, I'm just going to merge the PR since I think this is reasonable to handle separately.